### DISCUSSION TITLE
Record directly to file

### DISCUSSION THREAD

**Discussion by LwsBtlr** [ExistentialAudio/BlackHole#8] [Feature Requests]
Any chance we will be able to direct blackhole to simply dump the audio directly into a file instead of using another program to “record” the audio stream and then saving it when it’s done?

**Comment by masonkoh** [comment]
That would be a quite powerful feature that makes this program even more perfect. I strongly suggest you implement that.

**Comment by iurikothe** [comment]
This would be lovely.

I'd use to record just audio from meetings (via Skype or Browsers)

**Comment by adarshp** [comment]
This would be a very useful feature for us.

**Reply by adarshp** [reply]
Follow up - I've changed my stance and agree with https://github.com/ExistentialAudio/BlackHole/discussions/8#discussioncomment-1856163 - there are other tools that take care of this. We are using `ffmpeg` to record directly to file.

**Comment by simcc** [comment]
A nice upgrade to that would be a timemachine...so say the last _x_ minutes is constantly written to a file...a "print button" saves the file with date/time as name.

**Comment by derkiki** [comment]
Record to file would be awesome. Could this be done faster than realtime? (fake a very high sample rate)

**Comment by devinroth** [comment]
Probably not. Usually when there is a massive difference in sample rates (actual vs reported) apps don’t like it. At least Logic I know for sure throws an error.  But it’s a cool idea in theory.

Devin

**Comment by AlgyTaylor** [comment]
+1 for recording audio to a file.  Obviously it's one that's fairly easy to get around (open audacity or whatever), but it's such a frequent pair of actions to do together ....

**Comment by yoiang** [comment]
It may be helpful to understand how people are thinking this would work before a BlackHole interface exists or if it would have to wait until then. Would it be enough that as soon as a **BlackHole xch to File** device was added in the Audio MIDI Setup that it would immediately start dumping to a specific file and only stop if muted or removed?

**Comment by devinroth** [comment]
It doesn’t make sense to have it built inside the driver.

Devin

**Comment by GrampaPaul** [comment]
It might be nice.  I am not sure how helpful or if it would be helpful.  I downloaded and installed BlackHole  on my previous iMac with OS10.11.6 El Capitan.  It solved a problem I was having with recording while listening.    I since replaced my iMac with a newer iMac having Catalina and brought BlackHole into it with anticipation of using it again.  Soon after this was updated to BigSur.  Thus far, I have not found a need for BlackHole.Grampa Paul

**Comment by vin-cf** [comment]
+1 Would be interested in having this feature

**Comment by mrickard** [comment]
I'm not a contributor, but as a user of BlackHole, I'd be in favor of *not* adding stream-to-file to BH. I use BH to stream audio from Max/MSP to Vuo and OBS. I record by streaming to disk both from Max and from another free tool ([butt](https://danielnoethen.de/butt/)). There are plenty of options for capturing audio that don't require adding complexity to BH.

**Comment by MaikWaschfeld** [comment]
HI,

>  We are using ffmpeg to record directly to file.
> 

Any chance, you could share your terminal script, or whatever you use, to achieve this?


With kindest regards…
…Maik Waschfeld

(sent from my MBPro16,2)
also at ***@***.***>

**Reply by adarshp** [reply]
Sure thing @MaikWaschfeld ! 

Here's the invocation: https://github.com/ml4ai/tomcat/blob/6d4745ab0993926dad3a45a87ff2ce9f3dd4c215/tools/run_session#L92-L96

- The environment variable `FFMPEG_FMT_SYSTEM_AUDIO` is set here: https://github.com/ml4ai/tomcat/blob/6d4745ab0993926dad3a45a87ff2ce9f3dd4c215/tools/configuration_helpers#L124
- The variable `FFMPEG_INPUT_DEVICE_SYSTEM_AUDIO` is set here: https://github.com/ml4ai/tomcat/blob/6d4745ab0993926dad3a45a87ff2ce9f3dd4c215/tools/configuration_helpers#L214

**Comment by orchetect** [comment]
File recording would not and should not be a feature of the driver itself.

A possibility would be a front-end GUI that could initiate and terminate disk capture. But it's a redundant feature when there are countless free and paid applications that you can simply route the audio to and record on your own terms. BlackHole's strength is its simplicity and stability and it should focus on that.
